File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/6_653.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        a = set()

        def dfs(node):
            if node == None:
                return False 
            if (k - node.val) in a:
                return True 
            else:
                a.add(node.val)
            
            return dfs(node.left) or dfs(node.right) 
        
        return dfs(root)

File: Algorithms_and_Data_structures/ADS_2021/w4_BST/lab4/leetcode_additional/6_653.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> v;
    bool flag = false;

    bool findTarget(TreeNode* root, int k) {
        dfs(root, k);
        return flag;
    }
    void dfs(TreeNode* node, int k){
        if (node == NULL){
            return;
        }
        int y = k - node->val;
        dfs(node->left, k);

        if (binary_search(v.begin(), v.end(), y) == true){
            flag = true;
            return;
        }
        else{
            v.push_back(node->val);
        }

        dfs(node->right, k);

    }
};